Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm

```diff
@@ -832,65 +824,61 @@ static void registerLogHook()
-        Vector<uint8_t> privdata(std::span { msg->privdata, msg->privdata_sz });
 
-        logQueue()->dispatchWithQOS([logFormat = WTFMove(logFormat), logChannel = WTFMove(logChannel), logCategory = WTFMove(logCategory), type = type, buffer = WTFMove(buffer), privdata = WTFMove(privdata), qos] {
-            os_log_message_s msg;
-            memset(&msg, 0, sizeof(msg));
+        char* messageString = os_log_copy_message_string(msg);
+        if (messageString) {
+            std::span logString(byteCast<uint8_t>(messageString), strlen(messageString) + 1);
+            WebProcess::singleton().sendLogOnStream(WTFMove(logChannel), WTFMove(logCategory), WTFMove(logString), type);
+            free(messageString);
+        }
+    }).get());
 
-            msg.format = logFormat.data();
-            msg.buffer = buffer.data();
-            msg.buffer_sz = buffer.size();
-            msg.privdata = privdata.data();
-            msg.privdata_sz = privdata.size();
+    WTFSignpostIndirectLoggingEnabled = true;
+}
 
-            char* messageString = os_log_copy_message_string(&msg);
-            if (!messageString)
-                return;
-            std::span logStringIncludingNullTerminator(messageString, strlen(messageString) + 1);
+void WebProcess::setupLogStream()
+{
+    if (os_trace_get_mode() != OS_TRACE_MODE_OFF)
+        return;
 
-            auto connectionID = WebProcess::singleton().networkProcessConnectionID();
-            if (connectionID)
-                IPC::Connection::send(connectionID, Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(logChannel.spanIncludingNullTerminator(), logCategory.spanIncludingNullTerminator(), logStringIncludingNullTerminator, type, getpid()), 0, { }, qos);
+    static constexpr auto connectionBufferSizeLog2 = 21;
+    auto connectionPair = IPC::StreamClientConnection::create(connectionBufferSizeLog2, 1_s);
+    if (!connectionPair)
+        CRASH();
+    auto [streamConnection, serverHandle] = WTFMove(*connectionPair);
+    m_logStreamConnection = WTFMove(streamConnection);
+    m_logStreamConnection->open(*this);
 
-            free(messageString);
-        }, qos);
+    ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetupLogStream(getpid(), WTFMove(serverHandle), WebProcess::singleton().m_logStreamIdentifier), [] (IPC::Semaphore&& wakeUpSemaphore, IPC::Semaphore&& clientWaitSemaphore) {
+        WebProcess::singleton().m_logStreamConnection->setSemaphores(WTFMove(wakeUpSemaphore), WTFMove(clientWaitSemaphore));
+#if PLATFORM(IOS_FAMILY)
+        prewarmLogs();
+#endif
+        WebProcess::singleton().registerLogHook();
     });
+}
 
-    WTFSignpostIndirectLoggingEnabled = true;
+void WebProcess::sendLogOnStream(std::span<const uint8_t>&& logChannel, std::span<const uint8_t>&& logCategory, std::span<uint8_t>&& logString, os_log_type_t type)
+{
+    if (!m_logStreamConnection)
```

Comment:
Personally, I like this format in such cases:
```
if (m_logStreamConnection)
    m_logStreamConnection->send(Messages::LogStream::LogOnBehalfOfWebContent(logChannel, logCategory, logString, type), m_logStreamIdentifier);
```

Reply:
Also note elsewhere in this PR that you should have a RefPtr to the logStreamConnection *on the stack* before calling non-trivial functions on it, as per our smart pointer guidelines